lib/connectors/memory.js

```diff
@@ -250,6 +250,43 @@ Memory.prototype.updateOrCreate = function(model, data, options, callback) {
     }
   });
 };
+Memory.prototype.patchOrCreateWithWhere =
+  Memory.prototype.upsertWithWhere = function(model, where, data, options, callback) {
+    var self = this;
+    var element = Object.keys(where)[0];
+    var filter = { where: where };
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
+    }
+    var primaryKey = this.idName(model);
+    self.all(model, filter, options, function(err, instances) {
```

Comment:
changed the implementation to make it atomic. now using _findAllSkippingIncludes instead of self.all.